uwblueprint/starter-code-v2
# Add TypeScript UserService and AuthService implementations using MongoDB

## Notion ticket link

* Addresses the user service part of [Firebase authentication & user services using MongoDB (Mongoose)](https://www.notion.so/uwblueprintexecs/Firebase-authentication-user-services-using-MongoDB-Mongoose-1618340d0baa4549a8e38cd521c149c0)
* [Configure TS logger for backend](https://www.notion.so/uwblueprintexecs/Configure-TS-logger-for-backend-f4234bb4329143548896825310127acb)



## Implementation description
* Implemented methods defined in `IUserService`
* Defined MongoDB `User` model using Mongoose
* Integrated Firebase Admin SDK for auth related user management logic
* Set up logger using [winston](https://www.npmjs.com/package/winston), which is useful for keeping track of possible inconsistencies across Firebase and MongoDB user data (caused by rollback failure edge cases)



## Steps to test
1. Generate a Firebase service account private key
    * Go to our project in the [Firebase console](https://console.firebase.google.com)
    * "Project settings" > "Service accounts" > "Generate private key"
    * A file will be downloaded, copy it into `/backend/typescript/` and rename it **`firebaseServiceAccount.json`**
 2. Pull from secret manager **OR** add this line to the root `.env` file in the repo
 ```
 GOOGLE_APPLICATION_CREDENTIALS=firebaseServiceAccount.json
 ```
 3. Comment out the `frontend` and `py-backend` services in `docker-compose.yml`
 4. Run the application:
 ```
docker-compose up --build
```
5. Send requests to the test endpoints in `backend/typescript/server.ts` (note they are all GET requests). Suggested test plan:
    * Create some users by sending requests to `/create-user/:firstName/:lastName`.
    * Update some users sending requests to `/update-user/:userId`
    * Retrieve a user by `userId` by sending a request to `/get-user-by-id/:userId`
    * Retrieve a user by `email` by sending a request to `/get-user-by-id/:email`
    * Retrieve all users by sending a request to `/get-users`
    * Delete a user by `userId` by sending a request to `/delete-user-by-id/:userId`
    * Delete a user by `email` by sending a request to `/delete-user-by-email/:email`

Please check the data in MongoDB Atlas and the Firebase console as well.



## What should reviewers focus on?
* The most important file to look at is `backend/typescript/services/implementations/userService.ts`
* Error handling and "rollbacks": does the logic make sense, are there cleaner ways to do this?
* Concerns on data consistency - seems like an unavoidable problem when we use both Firebase and another data store, but let me know if you think of a better way!
* Am I using TypeScript features properly?

## Checklist
- [x] I have run the appropriate linter(s) (note: unused var error on `IAuthService` is expected)
